### PR TITLE
Add ability to jsonify Drop

### LINKED ISSUE #4668: "Liquid Exception: stack level too deep" on jsonify
###### What version of Jekyll are you using (`jekyll -v`)?

jekyll 3.0.3

###### What operating system are you using?

Darwin ***** 15.3.0 Darwin Kernel Version 15.3.0: Thu Dec 10 18:40:58 PST 2015; root:xnu-3248.30.4~1/RELEASE_X86_64 x86_64

###### What did you do?
(Please include the content causing the issue, any relevant configuration settings, and the command you ran)

```
$ jekyll build
```

in `_layouts/city.html` this line is present:

```
{{ page | jsonify }}
````

###### What did you expect to see?

A complete build.

###### What did you see instead?

```
Configuration file: _config.yml
    Server address: http://127.0.0.1:4000//
  Server running... press ctrl-c to stop.
      Regenerating: 1 file(s) changed at 2016-03-14 22:08:52 ...done in 1.435773 seconds.
      Regenerating: 1 file(s) changed at 2016-03-14 22:08:56
  Liquid Exception: stack level too deep in _layouts/city.html
```

`{{ page | jsonify }}` causes the exception. This used to work previously. I assume `jsonify`ing the `page` object breaks because something in the page object is of recursive nature or not "dumpable". `jsonify`ing individual keys in `page` (e.g. `{{ page.item | jsonify }}`) works.

### PR DESCRIPTION
**Pull request by stefanw**
Currently `{{ page | jsonify }}` returns a JSON string of the page content while a JSON representation of the pages data is more useful. This PR adds a `to_json` to the `Drop` class and dumps the YAML front matter data.

The Drop contains more than just a page's YAML front matter data. However, dumping the complete drop is apparently not possible due to the recursive nature of some objects (e.g. probably `previous` and `next`).

This PR would close #4668.

### CODE REVIEW

**Comment by stefanw** [comment]
I updated the original PR to reflect that only YAML front matter as present in the Drop's `fallback_data` is JSON dumped.

File: lib/jekyll/drops/drop.rb [hunk 1]
```diff
@@ -137,6 +137,7 @@ def inspect
         require 'json'
         JSON.pretty_generate to_h
       end
+      alias_method :to_json, :inspect
```
Comment (envygeeks) [inline lib/jekyll/drops/drop.rb]: * Use alias without symbolizations: `alias method other_method`
* Group `alias_method` at the bottom of the class rather than close to the method.
* Use alias instead of alias method.

File: test/test_document.rb [hunk 2]
```diff
@@ -44,6 +44,16 @@ def assert_equal_value(key, one, other)
       assert_equal "foo.bar", @document.data["whatever"]
     end
 
+    should "be able to jsonify its data" do
+      require 'json'
```
Comment (envygeeks) [inline test/test_document.rb]: * Do not require within a test.  This is bad practice, if the library is not available you're doing it wrong.

File: lib/jekyll/drops/drop.rb [hunk 3]
```diff
@@ -137,7 +137,16 @@ def inspect
...
-      alias_method :to_json, :inspect
+
+      # When a JSON representation of the drop is requested (e.g. through
+      # template filters) then only show the YAML front matter data.
+      # Other Jekyll objects might not be dumpable due to recursiveness.
+      #
+      # Returns a JSON dump of the YAML front matter data.
+      def to_json
+        require 'json'
+        JSON.generate fallback_data
```
Comment (envygeeks) [inline lib/jekyll/drops/drop.rb]: * Prefer to use `to_json` instead of `JSON.generate` unless you intend to use `JSON.pretty_generate`

**Comment by envygeeks** [comment]
Just a few comments after the bot did it's review but I'm :+1: this seems to be solid.

File: test/test_document.rb (hunk 2 above)
Comment (stefanw) [inline test/test_document.rb]: How should I do it right then? I'm not a Ruby person, sorry!

File: lib/jekyll/drops/drop.rb (hunk 1 above)
Comment (stefanw) [inline lib/jekyll/drops/drop.rb]: I mimicked the method aliasing done for the `to_h` method. But this is not part of the PR now anyway ;)

File: test/test_document.rb (hunk 2 above)
Comment (envygeeks) [inline test/test_document.rb]: In this case I would just add it to Jekyll's main require file because it's part of StdLib but if you want to be local I would require it at the top of formatters file you edited because it has no performance impact and doesn't slow anything down because most of the time it's required anyways by one thing or another.

Comment (parkr) [inline test/test_document.rb]: Sorry for the tone of the bot, @stefanw, in that first comment. I'm sure it didn't mean to tell you off quite so harshly.

You should be able to remove this `require`. From the looks of it, the method above `#to_json` uses `JSON` too, so it should already be required. Were you seeing it break before?

Comment (parkr) [inline test/test_document.rb]: Ah, I misspoke. It's required in `#inspect`. So it looks like you'll have to manually require it either in `lib/jekyll.rb` or as the first line of the body of this method.

Why don't you just omit certain keys? Maybe something like:

```ruby
keys.each_with_object({}) do |(key, _), result|
  result[key] = self[key] unless %w{previous next}.include? key
end
```

File: lib/jekyll/drops/drop.rb [hunk 4]
```diff
@@ -138,6 +138,15 @@ def inspect
         JSON.pretty_generate to_h
       end
 
+      # When a JSON representation of the drop is requested (e.g. through
+      # template filters) then only show the YAML front matter data.
+      # Other Jekyll objects might not be dumpable due to recursiveness.
+      #
+      # Returns a JSON dump of the YAML front matter data.
+      def to_json
+        fallback_data.to_json
```
Comment (parkr) [inline lib/jekyll/drops/drop.rb]: This feels incomplete to me... We definitely need more than just the fallback data to be jsonified. Ideally we just call `.inspect` on the values in `next` and `previous`.

**Comment by parkr** [comment]
@stefanw Still have time to work on this, or are you unable to complete this? I have a concern above I'd like to know your thoughts on.

**Comment by stefanw** [comment]
@parkr I could whitelist or blacklist keys to include in the dump, but I guess the best way would be to have [all variables listed in the documentation](https://jekyllrb.com/docs/variables/) to be available in the dump.

In addition to `page`, `site` and `paginator` should also be dumpable, I think. The question remains how to avoid dumping without loops. E.g. I could see `page.previous` just becoming a URL path in the dump.

Is that what you have in mind?

**Comment by parkr** [comment]
> The question remains how to avoid dumping without loops.

If we can identify `page.next` and `page.previous` as problematic, I think representing them via `inspect` is fine. We can start with those two properties and add more later if they come up.

**Comment by stefanw** [comment]
Representing via `inspect` results in this though:

```json
{
"next": "#<Jekyll::Document _cities/rochester.md collection=cities>",
"previous":"#<Jekyll::Document _cities/rennes.md collection=cities>",
"path":"_cities/rhein-neckar.md",
"...": "..."
}
```

Is that really useful in a JavaScript context?

**Comment by pathawks** [comment]
> Is that really useful in a JavaScript context?

Nope. A URL would be useful.

**Comment by stefanw** [comment]
Now I'm calling `url` on the previous/next values when they're not `nil`.

**Comment by parkr** [comment]
@stefanw @pathawks A URL is invalid, though. If you were using `jsonify` to try to understand what `page.next` was, they'd see a URL even though it's the Document. We should represent this some other way.

**Comment by stefanw** [comment]
I used a URL in the spirit of HATEOAS. If you're interested in what the next document is, you can send a request to that URL which then contains that info.

**Comment by parkr** [comment]
@stefanw That's fine in theory but not everyone is writing a REST-ful service so HATEOAS is not valid here. People often get confused if `page.next` is a document or a URL and this would further confuse them. It really must be a document itself. It can be limited, but it must be a document.

**Comment by parkr** [comment]
#4945 fixed this.
